Add TI J721e R5 and BeagleBone AI64 R5 initial support. And fix up current PINCTRL selection for other TI boards.

glneo opened the pull request:
This PR adds SoC support for TI J721e, specifically the R5 cores found in the MAIN domain (there is an R5 in the MCU domain we will add support for later). We also add board support for the Beaglebone AI64 which uses the J721e SoC.

TRM for J721e: https://www.ti.com/lit/zip/spruil1
BeagleBone AI-64: https://beagleboard.org/ai-64

Note: this is the continuation of the PR #59191 which went stale and was closed. Updates include HWMv2 and other small bug fixes. This is based on #71526 which is the first patch here (included here also so this series can still build).

slpp95prashanth committed:
soc: ti_k3: Add TI J721E SoC R5

Add initial SoC support for the TI J721E SoC series Cortex-R5 core.

TRM for J721e https://www.ti.com/lit/zip/spruil1
File: spruil1c.pdf

Signed-off-by: Prashanth S <slpp95prashanth@yahoo.com>
Signed-off-by: Andrew Davis <afd@ti.com>

slpp95prashanth committed:
boards: beaglebone_ai64: Add BeagleBone AI-64

Add initial BeagleBone AI-64 support.

BeagleBone AI-64: https://www.beagleboard.org/boards/beaglebone-ai-64

Signed-off-by: Prashanth S <slpp95prashanth@yahoo.com>
Signed-off-by: Andrew Davis <afd@ti.com>

Ayush1325 commented on boards/beagle/beaglebone_ai64/doc/assets/bbai_64.png:
Probably should be converted to webp

glneo replied:
Sure, will convert.

Ayush1325 commented on boards/beagle/beaglebone_ai64/doc/index.rst:
```diff
@@ -0,0 +1,90 @@
+Use the latest Linux image from https://beagleboard.org/latest-images to bootup
+the BBAI-64. This is required for loading zephyr.elf/binary using remoteproc.
+
+Steps to flash the image
```
You can probably create a script which can be called in `board.cmake` to allow `west flash` to work.

glneo replied:
This board doesn't really have flash, nor a good way to "flash" from a running system. I'll update this doc to explain better.

nordicjm commented on drivers/timer/Kconfig.omap_dm_timer:
```diff
@@ -0,0 +1,13 @@
+	select TICKLESS_CAPABLE
+	depends on DT_HAS_TI_AM654_DMTIMER_ENABLED
+	help
+		This module implements a kernel device driver for TI dual-mode timer. This
```
Kconfig help indent is 1x tab followed by 2x spaces

glneo replied:
Fixed

nordicjm commented on soc/ti/k3/j721e/Kconfig.soc:
```diff
@@ -0,0 +1,36 @@
+	help
+	  Enable support for the TI Jacinto J721E SoC series
+
+config SOC_SERIES_J721E_A72
+	bool
+	select SOC_SERIES_J721E
+	help
+	  Enable support for TI K3 J721e A72 Series
+
+config SOC_SERIES_J721E_R5
+	bool
+	select SOC_SERIES_J721E
+	help
+	  Enable support for TI K3 J721e R5 Series
```
these symbols to go, they are not series

glneo replied:
Sure, removed.

nordicjm commented on soc/ti/k3/j721e/Kconfig.soc:
```diff
@@ -0,0 +1,36 @@
+	  Enable support for TI K3 J721e R5 Series
+
+config SOC_J721E_A72
+	bool
+	select SOC_SERIES_J721E_A72
+
+config SOC_J721E_R5
+	bool
+	select SOC_SERIES_J721E_R5
```
select SOC_SERIES_J721E

glneo replied:
ACK

nordicjm commented on soc/ti/k3/j721e/Kconfig.soc:
```diff
@@ -0,0 +1,36 @@
+# SPDX-License-Identifier: Apache-2.0
+
+config SOC_SERIES_J721E
+	bool "TI Jacinto J721E SoC series"
```
```suggestion
	bool
```

glneo replied:
Removed.

nordicjm commented on soc/ti/k3/j721e/Kconfig.defconfig:
```diff
@@ -0,0 +1,41 @@
+	default y
+
+if !XIP
+	config FLASH_SIZE
+		default 0
+
+	config FLASH_BASE_ADDRESS
+		default 0
```
no extra indent

glneo replied:
Indent removed

nordicjm commented on soc/ti/k3/j721e/Kconfig:
```diff
@@ -0,0 +1,21 @@
+	select OPENAMP_RSC_TABLE
+	select UART_NS16550_ACCESS_WORD_ONLY
+
+config SOC_PART_NUMBER
+	default "J721e" if SOC_J721E_A72
+	default "J721e" if SOC_J721E_R5
```
`Kconfig.soc` as
```
config SOC_PART_NUMBER
	default "J721e" if SOC_J721E_A72 || SOC_J721E_R5
```

nordicjm commented on boards/beagle/beaglebone_ai64/Kconfig.beaglebone_ai64:
```diff
@@ -0,0 +1,8 @@
+# SPDX-License-Identifier: Apache-2.0
+
+config BOARD_BEAGLEBONE_AI64_J721E_R5
+	bool "BeagleBone AI-64 R5"
```
No selection

nordicjm commented on boards/beagle/beaglebone_ai64/Kconfig.beaglebone_ai64:
```diff
@@ -0,0 +1,8 @@
+#
+# SPDX-License-Identifier: Apache-2.0
+
+config BOARD_BEAGLEBONE_AI64_J721E_R5
```
wrong board name, no socs in here

nordicjm commented on boards/beagle/beaglebone_ai64/Kconfig.beaglebone_ai64:
```diff
@@ -0,0 +1,8 @@
+
+config BOARD_BEAGLEBONE_AI64_J721E_R5
+	bool "BeagleBone AI-64 R5"
+	select SOC_J721E_R5
```
select x if *use board name with socs*

glneo replied:
Will fix.

Also, I copied this from `boards/ti/sk_am62/Kconfig.sk_am62`, I guess that will need fixed the same way, I'll go do that too.

vaishnavachath commented on soc/ti/k3/j721e/CMakeLists.txt:
```diff
@@ -0,0 +1,21 @@
+zephyr_include_directories(.)
+
+if(CONFIG_SOC_J721E_A72)
+  set(SOC_LINKER_SCRIPT ${ZEPHYR_BASE}/include/zephyr/arch/arm64/scripts/linker.ld CACHE INTERNAL "")
```
Also there are few other places introducing A72 related config entries but those are not being unused, are those changes intended to be part of this PR?

glneo replied:
This series adds the SoC and just one board (AI64) using it, but I do use the A72 support for a different board (SK-TDA4VM) that I'm not in this PR. Should I drop the A72 support here and add it back when I push the next board PR, or is it fine to leave it to get ready for later? I have no preference.

dnltz replied:
If it's not too much work, it's probably better to remove code for the A72 from this PR. There is still too much missing to get Zephyr booting on these cores like MMU regions.

glneo replied:
Easy to drop, will remove the A72 stuff.

vaishnavachath commented on include/zephyr/drivers/timer/ti_dmtimer.h:
```diff
@@ -0,0 +1,158 @@
+#define TI_DM_TIMER_TCVR           (TI_DM_TIMER_BASE_ADDR + 0x64)
+#define TI_DM_TIMER_TOCR           (TI_DM_TIMER_BASE_ADDR + 0x68)
+#define TI_DM_TIMER_TOWR           (TI_DM_TIMER_BASE_ADDR + 0x6c)
+
```
Precalculating the register addresses for a single hardcoded instance seems not scalable especially when multiple timer instances are present, as discussed in the DM timer PR I will recommend fixing this.

glneo replied:
Sure, will go fix that in the DM timer PR (and rebase so the fix ends up here too).

gramsay0 commented on boards/beagle/beaglebone_ai64/doc/index.rst:
```diff
@@ -0,0 +1,105 @@
+* 16 KB data cache.
+* 64 KB TCM.
+
+Region Address Translation
```
Q: Other k3 SoCs configure RAT in soc.c, is that required here?

glneo replied:
It can be used, but is not required. For now we use the default RAT mapping.

gramsay0 commented on soc/ti/k3/j721e/Kconfig.soc:
```diff
@@ -0,0 +1,24 @@
+#
+# SPDX-License-Identifier: Apache-2.0
+
+config SOC_SERIES_J721E
```
It looks like there are more SoCs in this series (e.g. J7200), consider renaming to `SOC_SERIES_J7X` or similar

dnltz replied:
I agree, this should be more generic like we did for the am6x.

glneo replied:
I'm okay with more generic, but I don't think giving it a more generic names helps us here.

TI's K3 processor line naming scheme doesn't tend to work well with generic names and wildcards like "x". For instance AM62P and AM62A are very different parts, and there is also a AM62 part which is very different still from the other two. So AM62x would be an incorrect "generic" as it would seem to support 3 parts, which definitely do not have common M4 Zephyr support.

I'd like to do what DT folks do with compatible strings: use the full name of the first device that is compatible, then all later compatible devices simply re-use that name. Only when an incompatible device comes out do we add the next name. So here, J721e is the first part with R5 cores, J7200, J784s4, etc, would simply make use of it. Then J722s, which has different R5 support, would simply start the new line for which follow up compatible devices then use.

gramsay0 replied:
That's a great point, let's wait for the next SoC(s) and refactor if/when it makes sense

nordicjm replied:
Can you pick the right name now? It's better to have a soc named correctly at the start than to add it with a wrong name then break it entirely for any out of tree users that have started using it when they update after you've fixed it

gramsay0 reviewed (commented):
Looks pretty good, mostly questions on things I'm uncertain on

glneo commented on soc/ti/k3/j721e/Kconfig.defconfig:
```diff
@@ -0,0 +1,43 @@
+	default 512
+
+config SYS_CLOCK_HW_CYCLES_PER_SEC
+	default 19200000
```
Yes, the timer is feed from a different clock. We will describe that in DT when we add a proper clock driver for K3 (TISCI based) that way we do not need to hard-code it here.

The DT is also not correct here, this core runs at 1GHZ. That `clock-frequency` isn't used anyway, so I'll just remove that.

glneo commented on soc/ti/k3/j721e/Kconfig:
```diff
@@ -0,0 +1,20 @@
+	select VIM
+	select TI_DM_TIMER
+	select OPENAMP_RSC_TABLE
+	select UART_NS16550_ACCESS_WORD_ONLY
```
ACK

dnltz commented on boards/beagle/beaglebone_ai64/beaglebone_ai64_j721e_r5.yaml:
```diff
@@ -0,0 +1,17 @@
+name: BeagleBone-AI64 R5
+type: mcu
+arch: arm
+ram: 32
```
ATCM and BTCM should be 64.

glneo replied:
The BTCM is at a non-contiguous address from the ATCM. I've not found a good way without really messing with the linker files to use both. Do you know of some standard way to have the binary sections split across two memories like that?

If not then we only really have 32KB RAM usable (until I add the DDR section).

dnltz replied:
Do we need to add virtual memory on top of both sections? I'm not familiar with that too. Might be easier to use DDR and just ignore BTCM.

The documentation mentions 64kB as available memory. Maybe this should be 32, too?

glneo replied:
The documentation is combining both 32KB ATCM and 32KB BTCM.

I'm planning on going with DDR as the "main" memory for this core (which would mean 16MB as that is how much we carveout for each R5 core today from Linux). But to use that we will want to turn on caching (DDR is far away from the core) and add RAT support which is going to be added in the next PR. Wanted to keep this PR simple.

dnltz commented on soc/ti/k3/j721e/pinctrl_soc.h:
```diff
@@ -0,0 +1,40 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_SOC_ARM64_TI_K3_PINCTRL_SOC_H_
```
Wrong define here.

glneo replied:
This file is an exact copy of the one in the `soc/ti/k3/am64` directory. The pinctrl should be common for all K3 SoCs, maybe the better solution is to move this file up to the common area. I'll add a patch for that.

dnltz replied:
Ah, sorry. I read `AM64` instead of `ARM64`. This is obviously fine.

glneo replied:
Yeah, I miss-read that name all the time, but this was still wrong since this file is no longer in the `ARM64` directory either. I've fixed up the definition anyway.

glneo commented on boards/beagle/beaglebone_ai64/Kconfig.beaglebone_ai64:
```diff
@@ -0,0 +1,7 @@
+# Copyright (C) 2023 BeagleBoard.org Foundation
+# Copyright (C) 2023 S Prashanth
```
Much of this was based on their work back in 2023 in PR #71526. I've added my(TI) copyright with updated year on the files that I have changed and/or made copyrightable changes.

glneo commented on boards/beagle/beaglebone_ai64/beaglebone_ai64_j721e_r5.dts:
```diff
@@ -0,0 +1,61 @@
+&pinctrl {
+	uart2_tx_default: uart2_tx_default {
+		/* 0x1c is address of padconfig register of p8.34 and 14 is mux mode */
+		pinmux = <0x1c (PIN_OUTPUT | MUX_MODE_14)>;
```
Good idea, will update to use the existing `K3_PINMUX` macro.

vaishnavachath commented:
Adding DNM as the dependency is yet to be merged and the PR needs to drop the DM timer commits : https://github.com/zephyrproject-rtos/zephyr/pull/71526

gramsay0 commented on drivers/timer/ti_dmtimer.c:
```diff
@@ -0,0 +1,167 @@
+	sys_write32(reg_val, reg);
+}
+
+void ti_dmtimer_isr(void *data)
```
`static`

glneo replied:
ACK, will fix that in the DM timer series.

nordicjm commented on soc/ti/k3/j721e/Kconfig:
```diff
@@ -0,0 +1,15 @@
+	select OPENAMP_RSC_TABLE
+	select UART_NS16550_ACCESS_WORD_ONLY if UART_NS16550
+
+config SOC_PART_NUMBER
+	default "J721e" if SOC_J721E_R5
```
this should be in `Kconfig.soc`

glneo replied:
Moved.

nordicjm commented on soc/ti/k3/j721e/Kconfig.soc:
```diff
@@ -0,0 +1,22 @@
+config SOC_SERIES_J721E
+	bool
+	select SOC_FAMILY_TI_K3
+	help
+	  Enable support for the TI J721E SoC series
```
```suggestion
```

glneo replied:
ACK

nordicjm commented on boards/beagle/beaglebone_ai64/doc/index.rst:
```diff
@@ -0,0 +1,105 @@
+
+Hardware
+********
+The BeagleBone AI-64 is powered by TI J721E SoC, which has three domains (Main,
```
minor nit. newline gaps after headings would be good

nordicjm commented on boards/beagle/beaglebone_ai64/doc/index.rst:
```diff
@@ -0,0 +1,105 @@
+
+Other hardwares features are currently not supported.
+
+The default configuration can be found in the defconfig file.
```
```suggestion
```

glneo replied:
Removed.

vaishnavachath commented:
@glneo Now that the DM timer series is merged, removing DNM, can you please drop those commits from here? Also this is going to be accepted only after 3.7.0 release.

glneo commented:
> @glneo Now that the DM timer series is merged, removing DNM, can you please drop those commits from here? Also this is going to be accepted only after 3.7.0 release.

As we have some time, I've made a couple additional changes. Since we have AM6x devices that also have the same R5 cores, making a new SoC series just for J7x makes it harder to later add the AM6x R5 devices. I'm just going to add this SoC to the existing SoC series. (I'd argue that K3 is the series and so having a AM6x "series" is unneeded also, but that is a cleanup for later).

Also, since the J721e SoC has 6(!) R5 cores, the naming needed updating so we can specify which we are targeting. When all core support is added `soc.yml` will look like:

```
    - name: j721e
      cpuclusters:
      - name: main/r5f0_0
      - name: main/r5f0_1
      - name: main/r5f1_0
      - name: main/r5f1_1
      - name: mcu/r5f0_0
      - name: mcu/r5f0_1
```
These names match what we call them in Linux also.

tejlmand commented on soc/ti/k3/soc.yml:
```diff
@@ -10,3 +10,6 @@ family:
       - name: m4
+    - name: j721e
+      cpuclusters:
+      - name: main/r5f0_0
```
blocking comment.

Looking at the datasheet for j721e I do see the use of domains, and that there are two domains, main and mcu, each having a number of cores.

So I understand the wish / logic in representing this by exposing the domains in the cluster name, unfortunately the `/` is also used as separator for the board qualifier and thus i'm concerned how underlying implementation will handle an extra `/` inside the CPU cluster. Especially as there can be multiple levels of board variants, and well as a future enhancements, see https://github.com/zephyrproject-rtos/zephyr/issues/69546, could suddenly be impacted.

Currently my expectation is that the `r5f0_0` probably will be treated as a board variant and thus work, but before approving I would like to take an extra look before deciding if this construct can be allowed, or if an extra field should be introduced for CPU cluster.

Alternative could be to name the cpu cluster as: `main_rf50_0`

glneo replied:
So the concern is around how to parse out which part of the following is the board:

`-b <board>/<board variant>/<soc>/<domain>/<core>`

when there are variants/revisions of a board?

Not sure why the board variant/revision would be put in the target string that way to begin with. It doesn't follow the normal pattern of the `/` separator meaning a nesting (i.e. contains one or more of x). Boards contain a number of SoCs, they do not *contain* a number of revisions.. Our SoCs do contain a number of domains which contain a number of cores, that matches perfectly with the existing convention.

I'd suggest then board revision simply be backed into the board string (e.g. board_deluxe_edition_revc/soc/etc..). That way, should some some variant of a board contain a different set of SoCs than the original, it could be represented correctly.

If the `/` does cause end up causing some issue, and we do not want to add sub-fields to CPU cluster, then main_rf50_0` is also fine with me if needed also.

glneo replied:
@tejlmand , gentle ping. Any comment on the above? This seems to be the last outstanding item on this PR

glneo replied:
To be safe I've just changed the name to `main_rf50_0`, should resolve your concern.

glneo commented on soc/ti/k3/am6x/CMakeLists.txt:
```diff
@@ -18,4 +18,15 @@ elseif(CONFIG_SOC_SERIES_AM6X_M4)
+elseif(CONFIG_SOC_SERIES_AM6X_R5)
+  zephyr_sources(r5/soc.c)
+
+  zephyr_include_directories(r5/)
```
ACK, will fix.

jadonk commented on boards/beagle/beaglebone_ai64/doc/index.rst:
```diff
@@ -0,0 +1,112 @@
+Hardware
+********
+
+The BeagleBone AI-64 is powered by TI J721E SoC, which has three domains (Main,
```
Branding thing.... Don't put "The" before "BeagleBone AI-64". See https://www.beagleboard.org/brand-use.

glneo replied:
Sure, will fix.

jadonk commented on boards/beagle/beaglebone_ai64/doc/index.rst:
```diff
@@ -0,0 +1,112 @@
+| DDR0       | 0xA2000000   | 0xA3000000   | 0xA4000000   | 0xA5000000   | 1MB    |
++------------+--------------+--------------+--------------+--------------+--------+
+| DDR1       | 0xA2100000   | 0xA3000000   | 0xA4100000   | 0xA5000000   | 15MB   |
++------------+--------------+--------------+--------------+--------------+--------+
```
Which of these are defined by the hardware and which are defined by convention and the Linux kernel device tree? If something can be changed, I'd like to see that noted along with where it is specified in both this code as well as Linux device tree or other outside code.

glneo replied:
ATCM/BTCM are fixed, DDR carve-out regions can be changed. I'll update the docs to note this and explain where the changes would need be made.

jadonk replied:
Thanks for making it clearer.

jadonk commented on boards/beagle/beaglebone_ai64/doc/index.rst:
```diff
@@ -0,0 +1,112 @@
+| ``cp build/zephyr/zephyr.elf /lib/firmware/``
+|
+| Ensure the Core is not running.
+| ``echo stop > /sys/class/remoteproc/remoteproc18/state``
```
The `remoteproc18` is not hard-coded, as in, it is assigned based on configuration order during boot and might change between Linux kernel, device-tree and other Linux system changes. We've been trying to setup a set of udev rules that provide symlinks to absolute paths based on the hardware core. Please try to use those to provide documentation that is accurate over a longer term.

glneo replied:
I'm assuming [these are the latest names](https://github.com/beagleboard/repos-arm64/blob/main/bb-customizations/suite/trixie/debian/86-remoteproc-noroot.rules)? So that would mean for this `j7-main-r5f0_0`. Seems reasonable, I'll update.

(I do have a couple issues with the names though, but I can take that up later on the BeagleBoard forums/discord)

Ayush1325 replied:
Seems good now.

jadonk commented on boards/beagle/beaglebone_ai64/doc/index.rst:
```diff
@@ -0,0 +1,112 @@
+
+The Zephyr on BeagleBone AI-64 J721E Cortex R5 uses UART 2 (Rx p8.22, Tx p8.34)
+as console.
+
```
Do you have a roadmap to create an RPMsg-based console in the future? An entirely in-system development solution seems desirable to me. I believe OpenOCD is a solution for debug, but simple console messages should also be possible.

glneo replied:
Yes, we will need to move to the RPMsg-based console when we can as there are not enough physical pins/UARTs for all the remote cores on this SoC to get their own (3 DSPs, 6 R5s, etc..).

The RPMsg console requires IPC/MBOX support which is still being added[0]. This physical UART use should be removed by default after we get that support. (Or I could drop its use now, but it does act as a good initial example IMHO)

[0] https://github.com/zephyrproject-rtos/zephyr/pull/71241

jadonk replied:
I agree it is useful and a fine default until the RPMsg console is available. When it is there, switching the default would be good and make this connection an alternative.

cfriedt replied:
+1 to RPMsg based console. I've heard there may be recent improvements to virtio that could improve that too.

Kind of similar talk

https://youtu.be/kjoQ2Gidbuo?si=wWyVNP0ipjlkY7DM

jadonk commented on boards/beagle/beaglebone_ai64/doc/index.rst:
```diff
@@ -0,0 +1,112 @@
+References
+**********
+
+* `BeagleBone AI-64 Homepage <https://beagleboard.org/ai-64>`_
```
Use the URL of the live page (https://www.beagleboard.org/boards/beaglebone-ai-64) or the vanity URL (https://beaglebone.ai/64).

glneo replied:
Sure, updated.

jadonk reviewed (changes requested):
I look forward to seeing the resolution on the cpuclusters naming convention. Otherwise, I have minor branding nitpicks and the overall structure looks good to me.

Ayush1325 reviewed (dismissed):
LGTM. We probably want to wait for @tejlmand regarding the blocking comment.

Ayush1325 commented:
> > LGTM. We probably want to wait for @tejlmand regarding the blocking comment.
> 
> I replied to that comment months ago, then pinged it again 4 weeks back, still no update from @tejlmand. I'd hope we don't block indefinitely. Worst case we take this now and then have to later change the cluster name from `main/r5f0_0` to `main_r5f0_0` which wouldn't be a big deal at all.

I guess @andyross can decide what to do here. 
I would also prefer to get this merged now since it is also blocking beagley-ai support. And I personally also prefer the current naming.

vaishnavachath reviewed (changes requested):
@glneo Looking at the terminology again it looks like @tejlmand 's comment is valid and main_r5f0_0 is a better choice :
See https://docs.zephyrproject.org/latest/hardware/porting/board_porting.html#board-terminology and https://docs.zephyrproject.org/latest/glossary.html#term-board-qualifiers

What happens when you have secure/non-secure variants come into picture:
```
beaglebone_ai64/j721e/main/r5f0_0
```
I would recommend keeping that variant field open for future , when we might need them (secure/non secure , lockstep/non-lockstep .etc) , given the details are documented clearly, let's not deviate unless you see any benefit having the voltage domain information separately.

Sorry for the last minute comment, but my recommendation is to fix this if there are no concerns, as this platform addition is going to be used as a reference for multiple other platforms in future.

glneo commented:
Sure, if we think it makes sense to change the name will change it.

wmamills commented:
@glneo there is a message about clang-format that says it has not been addressed.  Did you look at that?
https://github.com/zephyrproject-rtos/zephyr/actions/runs/11060123536/job/30729987820?pr=71527

Perhaps this is one thing holding up this PR

wmamills commented:
@microbuilder @andyross This PR looks like it should be assigned to Kevin instead of Andy.  Do you guys agree?

microbuilder commented:
> @microbuilder @andyross This PR looks like it should be assigned to Kevin instead of Andy. Do you guys agree?

Hey Bill. Not sure there's anything here I'd be particularly expert at. The basic PR looks fine to me, though.

@carlescufi Can you ping Torsten on this one?

kartben commented on boards/beagle/beaglebone_ai64/doc/index.rst:
```diff
@@ -0,0 +1,121 @@
+
+The example shows how to load an image on Cortex R5FSS0_CORE0 on J721e.
+
+| Copy Zephyr image to the /lib/firmware/ directory.
```
Would be great to first explain how to *build* the image? :)

glneo replied:
Good idea, added.

jan-kiszka commented on soc/ti/k3/am6x/Kconfig.defconfig:
```diff
@@ -16,11 +19,13 @@ config NUM_IRQS
 	int
 	default 400000000 if SOC_SERIES_AM6X_M4
 	default 200000000 if SOC_SERIES_AM6X_A53
+	default 19200000 if SOC_SERIES_AM6X_R5
```
Seems to me as if this is not generally true. For the AM65x, 200000000 give accurate timing, 19200000 is way too fast there.

glneo replied:
As this will change for different SoCs, perhaps this info belongs in the SoCs DT? I've seen something like this done elsewhere:

```
config SYS_CLOCK_HW_CYCLES_PER_SEC
	default $(dt_node_int_prop_int,/cpus/cpu@0,clock-frequency)
```

jan-kiszka replied:
Also the J721e should be affected by https://github.com/zephyrproject-rtos/zephyr/blob/8068cb2567508d760e71f04042f0301fdd081403/drivers/timer/ti_dmtimer.c#L26
Or do you have `CONFIG_TIMER_READS_ITS_FREQUENCY_AT_RUNTIME` enabled?

glneo replied:
The setting here will depend on which clock is feeding the DM timer you are using. It is not the arch timer like with M4. Since the clock is delectable at runtime, once we get TI-SCI clock support in we will enable `CONFIG_TIMER_READS_ITS_FREQUENCY_AT_RUNTIME`.

gramsay0 replied:
I would avoid `CONFIG_TIMER_READS_ITS_FREQUENCY_AT_RUNTIME` unless really necessary. Comment from a kernel maintainer:

![Screenshot_20241005-172133](https://github.com/user-attachments/assets/c4b6c444-7f39-4bfa-96c5-bf679a563279)

jan-kiszka replied:
So, what to do with this setting for now? Leave it and revisit when adding more SoCs under `SOC_SERIES_AM6X_R5`?

glneo replied:
I'd like to just revisit this when we add SOC_AM6548_R5.

vaishnavachath commented:
@carlescufi @nashif @cfriedt Can we please merge this PR?

cfriedt commented:
@vaishnavachath - it's in the queue and will get pulled soon.

vaishnavachath commented:
Thank you @cfriedt

kartben commented:
> @vaishnavachath - it's in the queue and will get pulled soon.
> 
> https://merge-list.zephyrproject.io/

Not until assignee approves, no :)

kartben commented:
This should be assigned to @jadonk as Beagle maintainer but this doesn't seem possible. @jadonk can you confirm you are a collaborator in the GitHub org? Looks like you might not based on the grey, not green, checkmarks next to your reviews?

nordicjm commented on boards/beagle/beaglebone_ai64/beaglebone_ai64_j721e_main_r5f0_0_defconfig:
```diff
@@ -0,0 +1,18 @@
+# Zephyr Kernel Configuration
+CONFIG_XIP=n
+
+# Enable Pinctrl
+CONFIG_PINCTRL=y
```
Not allowed

Ayush1325 replied:
@glneo You need to select pinctrl at `soc/ti/k3/am6x/Kconfig`.
See https://github.com/zephyrproject-rtos/zephyr/pull/64718#discussion_r1734288793

glneo replied:
Seems PINCTRL was already set to default y in `soc/ti/k3/am6x/Kconfig.defconfig` so this line wasn't needed anyway, removed.

nordicjm replied:
> Seems PINCTRL was already set to default y in soc/ti/k3/am6x/Kconfig.defconfig so this line wasn't needed anyway, removed.

Even worse, that needs to be removed and it done properly

Ayush1325 replied:
@nordicjm It should be removed, but should that really be part of this PR? I think it might be sufficient to just add `select PINCTRL` in `SOC_SERIES_AM6X_R5` and leave cleaning up am6x in a separate PR.

nordicjm replied:
It's wrong and needs to be done properly

Ayush1325 replied:
> It's wrong and needs to be done properly

It became the wrong way to do it at some point after am6x support landed.

This commit only adds AM6X_R5 support, and thus, yes, it should be done properly for anything new. As such `select PINCTRL` should be added in `SOC_SERIES_AM6X_R5`.

But I don't see why fixing all the other AM6x boards should be part of this PR.

nordicjm replied:
This is adding a new SoC, it needs to add the selections properly, right now the selections are wrong. Will doing it properly for this SoC fix the other SoC? Yes

Ayush1325 replied:
> This is adding a new SoC, it needs to add the selections properly, right now the selections are wrong. Will doing it properly for this SoC fix the other SoC? Yes

Ok, let me clarify something. For me, doing it properly would mean 
1. Removing `CONFIG_PINCTRL=y` selection from `soc/ti/k3/am6x/Kconfig.defconfig`.
2. Adding `select PINCTRL` to `SOC_SERIES_AM6X_A53`, `SOC_SERIES_AM6X_M4` 
3. Removing `CONFIG_PINCTRL=y` from any board using these SOCs

Are you referring to only doing 1 and 2 in the PR? Or should 3 also be handled here? Because this PR actually only adds `AM6X_R5` and does not touch `AM6X_A53` and `AM6X_M4`.

nordicjm replied:
All 3. Honestly the time with all these messages is more than the total time it would take to fix it and update the PR

Ayush1325 replied:
Also, it might be good to link this issue here for more context on what @nordicjm is referring to: https://github.com/zephyrproject-rtos/zephyr/issues/78619

Ayush1325 replied:
> All 3. Honestly the time with all these messages is more than the total time it would take to fix it and update the PR

I mean sure, if someone already knew what you wanted them to do. The thing is that the number of boards that correctly do it is in the minority right now, so it would be great if someone linked https://github.com/zephyrproject-rtos/zephyr/issues/78619 . I only found out that issue existed today.

In my Beagleplay PR, the only board that followed the pattern was esp32, and that was basically it. So I am trying to solidly define what needs to be done to get this merged.

nordicjm replied:
There's over 600 boards in zephyr, that PR links the boards that do it wrong in the first post and the SoCs that do it wrong in the bottom post (which I'm thankful for you for mentioning earlier as I was not aware SoCs were doing it wrong also), there's about 150 or so checkboxes in that issue, so that's a minority of the 600+ supported boards. The way to do it for board and SoC is the same, SoC drivers select it, if a board needs it because it has a custom .c file (this is the only valid reason a board should select it) then the board Kconfig file (note: `Kconfig` not `Kconfig.<board>`) selects it in the same way

glneo replied:
Okay, will add a patch fixing the PINCTRL selection for other boards and SoCs as part of this series too.

gmarull commented on boards/beagle/beaglebone_ai64/beaglebone_ai64_j721e_main_r5f0_0.dts:
```diff
@@ -0,0 +1,69 @@
+	status = "okay";
+};
+
+&pinctrl {
+	uart2_tx_default: uart2_tx_default {
+		/* 0x1c is address of padconfig register of p8.34 and 14 is mux mode */
+		pinmux = <K3_PINMUX(0x1c, PIN_OUTPUT, MUX_MODE_14)>;
+	};
+
+	uart2_rx_default: uart2_rx_default {
+		/* 0x14 is address of padconfig register of p8.22 and 14 is mux mode */
+		pinmux = <K3_PINMUX(0x14, PIN_INPUT, MUX_MODE_14)>;
+	};
+};
```
please, move pinctrl settings to `board-pinctrl.dtsi`

glneo replied:
Done.

glneo committed:
soc: ti: k3: Select PINCTRL in UART driver not Kconfig.defconfig

The default configuration for PINCTRL should not be set with
the other default configurations in .defconfig, instead select
a default value as part of defining the UART driver.

Signed-off-by: Andrew Davis <afd@ti.com>

nordicjm commented on soc/ti/k3/am6x/Kconfig.soc:
```diff
@@ -29,9 +35,14 @@ config SOC_AM6442_M4
 	bool
 	select SOC_SERIES_AM6X_M4
 
+config SOC_J721E_R5
+	bool
+	select SOC_SERIES_AM6X_R5
+
 config SOC_SERIES
 	default "am6x" if SOC_SERIES_AM6X
 
 config SOC
 	default "am6234" if SOC_AM6234_M4 || SOC_AM6234_A53
 	default "am6442" if SOC_AM6442_M4
+	default "j721e" if SOC_J721E_R5
```
only just noticed the difference here, these need to match with the cpucluster from the soc.yml file so if that's `main_r5f0_0` then this need to be `config SOC_J721E_MAIN_R5F0_0` not sure if @tejlmand has a comment on this

glneo replied:
Sure, updated.

nordicjm commented on soc/ti/k3/am6x/Kconfig:
```diff
@@ -1,6 +1,9 @@
 # Copyright (c) 2023 Enphase Energy
 # SPDX-License-Identifier: Apache-2.0
 
+config SOC_SERIES_AM6X
+	select PINCTRL
```
this is not correct, the **drivers** need to select it

Ayush1325 replied:
@nordicjm I am a bit confused here, since I think beagleplay and beagleconnect freedom are also currently doing this wrong. There is `select PINCTRL` in soc Kconfig: https://github.com/zephyrproject-rtos/zephyr/blob/4793e00ae803ae6b7948e4c110f7ceef11821ef5/soc/ti/simplelink/cc13x2x7_cc26x2x7/Kconfig#L18

From your comment it seems that drivers are supposed to do this. However, where should pinctrl be enabled in case the board initialization needs it ?(the antenna switch in bcf and play case: https://github.com/zephyrproject-rtos/zephyr/blob/main/boards/beagle/beagleconnect_freedom/board_antenna.c)

glneo replied:
We use the PINCTRL for the `uart_ns16550.c` driver, but that driver doesn't select it, should I go and fix that up everywhere?

glneo replied:
That won't work either, as some users of `uart_ns16550.c` need PINCTRL and others do not. We can only know at the board level if pinmux will be needed. Not sure I know a good solution.

nordicjm replied:
`select PINCTRL if <soc_name>`

glneo replied:
This would need added for every current K3 <soc_name>, and then updated for every new one added, sounds like constant churn in the serial driver kconfig. Having it selected by the symbol `UART_NS16550_TI_K3` which is in `Kconfig.ns16550` and already selected by all the K3 SoCs seems to have the same effect (all K3 SoCs have pinmux) without listing each and every SoC in each and every driver's kconfig.

nordicjm replied:
Except if a user disables the UART, you're forcing on an option that is not needed because your dependencies are not correct

glneo replied:
If the user disables `SERIAL` then `UART_NS16550_TI_K3` will also not be enabled, and so `PINCTRL` would not be selected.

glneo replied:
BTW I just wanted to note that I do agree with what you are saying here.

If we want to do the right thing in all cases, then one must also note that `PINCTRL` is really only needed on boards that don't use the default pinmux settings for the SoC (all SoC pins have a default mux setting, if the board uses those then `PINCTRL` would also not be needed). This can only be known at the board level. But also as you note, only some drivers support `PINCTRL` in the first place (all should at some point, or it should be moved to a higher level like in Linux and done for drivers automatically in the `default` case).

So we only select `PINCTRL` if 3 things are true, the board's DT has a pinmux node associated with a driver, that driver uses pinmux, and that driver is enabled. If any of these or not true for all drivers used by a board, only then can `PINCTRL` be not selected.

I'm thinking there should be some DT parsing magic that can help with this. For all driver DT nodes that contain an enabled pinmux property we define `DT_HAS_<DRIVER>_PINMUX_ENABLED`. Then in Kconfig for drivers we can do this:

```
config <DRIVER>
	depends on DT_HAS_<DRIVER>_ENABLED
	select PINCTRL if DT_HAS_<DRIVER>_PINMUX_ENABLED
```

gmarull replied:
PINCTRL needs to be selected if the associated code makes use of the pinctrl API, that’s all you need to take care of.

jadonk commented on boards/beagle/beaglebone_ai64/beaglebone_ai64_j721e_main_r5f0_0_defconfig:
```diff
@@ -0,0 +1,15 @@
+# SPDX-License-Identifier: Apache-2.0
+
+# Zephyr Kernel Configuration
+CONFIG_XIP=n
```
Why is this not XIP? Is there some relocation happening?

glneo replied:
Kinda, we don't have FLASH memory to execute from, we always run out of RAM. Setting XIP causes some other issues as we do not have a hard-coded flash memory region. At some point we could consider the location Linux loads the binary to as the "flash" region, but that would need other refactoring and since we are loaded into memory that is not ROM we don't get much from XIP being set.

glneo commented on boards/beagle/beaglebone_ai64/beaglebone_ai64_j721e_main_r5f0_0.dts:
```diff
@@ -0,0 +1,57 @@
+#include <freq.h>
+
+/ {
+	model = "BeagleBone-AI64";
```
Will fix.

jadonk commented:
@nordicjm @gmarull have you reviewed since latest push? Do we have a good pattern for PINCTRL? I'll be working with @Ayush1325 and all to get fixes for other Beagles.

nordicjm commented:
> @nordicjm @gmarull have you reviewed since latest push? Do we have a good pattern for PINCTRL? I'll be working with @Ayush1325 and all to get fixes for other Beagles.

Drivers that need it must select it, it must not be in any defconfig files. If a board needs it for a board's .c file then it must be selected by the board `Kconfig` file

jadonk commented on soc/ti/k3/am6x/Kconfig.soc:
```diff
@@ -29,9 +35,14 @@ config SOC_AM6442_M4
 config SOC
 	default "am6234" if SOC_AM6234_M4 || SOC_AM6234_A53
 	default "am6442" if SOC_AM6442_M4
+	default "j721e" if SOC_J721E_MAIN_R5F0_0
```
This seems inconsistent with soc/ti/k3/am6x/Kconfig SOC_PART_NUMBER. Should it be leading caps, all caps or lower-case? Why would the SOC_PART_NUMBER be different than SOC?

glneo replied:
For `SOC` I found the following in the porting guide:

```
config SOC
    default "SoC name" if SOC_<SOC_NAME>

Notice that ``SOC_NAME`` is a pure upper case version of the SoC name.

Notice that the string value must match the SoC name used in the :file:`soc.yml` file.
```

To me that seems to indicate all lower to match what we do in soc.yml.

For `SOC_PART_NUMBER`, to be honest I have no clue, I don't see it used anywhere either, I just added it based on what the other TI SoCs had done, but maybe they didn't need it either?

nordicjm replied:
`SOC` must be lower, `SOC_PART_NUMBER` does not exist, i.e. it's a custom field, it has no relation to anything and is not used by anything, it can be safely removed unless you have HAL code using it

jadonk reviewed (approved):
Please read my question on SOC vs. SOC_PART_NUMBER -- not sure why they are different. This isn't a gate to my approval, just a query.

jadonk commented:
@gmarull  it seems the updates were made. Any changes remaining?
